File: scripts/split-admin-html.js
#!/usr/bin/env node
/**
 * Разбивает index.html на skeleton + pages/*.html
 *
 * Анализирует файл, находит все <div x-show="page === 'xxx'">,
 * вырезает их в отдельные файлы, остальное — skeleton.
 *
 * Запуск: node scripts/split-admin-html.js
 */

const fs = require('fs');
const path = require('path');

const htmlPath = process.argv[2] || 'admin-dashboard/internal/server/static/index.html';
const pagesDir = path.join(path.dirname(htmlPath), 'pages');

const html = fs.readFileSync(htmlPath, 'utf8');
const lines = html.split('\n');

// ── 1. Find all page sections ──
// Pattern: <!-- PageName -->\n<div x-show="page === 'xxx'" ...> ... </div>\n<!-- NextPage -->
const pageStartRe = /^\s*<div\s+x-show="page\s*===\s*'(\w+)'/;
const pageEndRe = /^\s*<\/div>/;

const pages = {}; // { name: { startLine, endLine, content: [] } }
let currentPage = null;
let depth = 0;

for (let i = 0; i < lines.length; i++) {
  const l = lines[i];

  // Check for page start
  const m = l.match(pageStartRe);
  if (m) {
    currentPage = { name: m[1], startLine: i, content: [l] };
    depth = 1;
    continue;
  }

  if (currentPage) {
    currentPage.content.push(l);

    // Track div nesting
    const openCount = (l.match(/<div[^>]*>/g) || []).length;
    const closeCount = (l.match(/<\/div>/g) || []).length;
    depth += openCount - closeCount;

    if (depth <= 0) {
      // End of this page
      currentPage.endLine = i;
      pages[currentPage.name] = currentPage;
      currentPage = null;
      depth = 0;
    }
  }
}

console.log(`Found ${Object.keys(pages).length} pages:`);
Object.entries(pages).forEach(([name, p]) => {
  console.log(`  ${name}: lines ${p.startLine}-${p.endLine} (${p.content.length} lines)`);
});

// ── 2. Write page files ──
fs.mkdirSync(pagesDir, { recursive: true });
const pageNames = Object.keys(pages).sort((a, b) => pages[a].startLine - pages[b].startLine);

pageNames.forEach(name => {
  const p = pages[name];
  const content = p.content.join('\n');
  fs.writeFileSync(path.join(pagesDir, `${name}.html`), content + '\n');
  console.log(`  wrote pages/${name}.html (${content.split('\n').length} lines)`);
});

// ── 3. Build skeleton ──
// Take all lines NOT in any page section
const pageLines = new Set();
pageNames.forEach(name => {
  const p = pages[name];
  for (let i = p.startLine; i <= p.endLine; i++) {
    pageLines.add(i);
  }
});

const skeletonLines = [];
let lastMarkerLine = -1;

for (let i = 0; i < lines.length; i++) {
  if (pageLines.has(i)) {
    // Check if we need to insert a marker
    if (lastMarkerLine !== i - 1) {
      // Find which page ends here and insert marker right before the next non-page line
      // We'll handle markers below
    }
    continue;
  }

  // Before writing this line, check if we just finished a page
  // and need to insert a marker
  let inserted = false;
  for (const name of pageNames) {
    const p = pages[name];
    if (p.endLine === i - 1) {
      skeletonLines.push(`<!--PAGE:${name}-->`);
      inserted = true;
    }
  }
  skeletonLines.push(lines[i]);
  lastMarkerLine = i;
}

// Also add markers for pages at very end
for (const name of pageNames) {
  const p = pages[name];
  if (p.endLine >= lines.length - 1) {
    skeletonLines.push(`<!--PAGE:${name}-->`);
  }
}

const skeleton = skeletonLines.join('\n');
fs.writeFileSync(htmlPath, skeleton);
console.log(`\nWrote skeleton index.html (${skeletonLines.length} lines, was ${lines.length})`);

// ── 4. Report ──
console.log('\nDone! Summary:');
console.log(`  skeleton: index.html (${skeletonLines.length} lines)`);
pageNames.forEach(name => {
  const f = path.join(pagesDir, `${name}.html`);
  const size = fs.statSync(f).size;
  console.log(`  pages/${name}.html: ${size} bytes`);
});

File: infra/scripts/split-admin-html.py
#!/usr/bin/env python3
"""
Robustly split admin-dashboard index.html into skeleton + pages/*.html.

Properly tracks <div> nesting by counting ALL occurrences of <div and </div> per line,
including multiple tags on the same line.
"""
import re, os

html_path = 'admin-dashboard/internal/server/static/index.html'
pages_dir = os.path.join(os.path.dirname(html_path), 'pages')

with open(html_path) as f:
    html = f.read()

lines = html.split('\n')

PAGE_RE = re.compile(r'<div\s+x-show="page\s*===\s*\'(\w+)\'"')

# Find exact line ranges for each page section by tracking div nesting
pages = {}

i = 0
while i < len(lines):
    line = lines[i]
    m = PAGE_RE.search(line)
    if m:
        name = m.group(1)
        start = i

        # Count depth: opening div contributes to depth
        depth = 0
        depth += line.count('<div') - line.count('</div>')

        # If there's an opening div on the first line (should always be at least 1)
        # scan until depth returns to 0
        j = i + 1
        while j < len(lines) and depth > 0:
            depth += lines[j].count('<div') - lines[j].count('</div>')
            j += 1

        end = j - 1
        pages[name] = {'start': start, 'end': end, 'name': name}
        print(f"  {name}: lines {start+1}-{end+1} ({end-start+1} lines), depth={depth}")
        i = end + 1
    else:
        i += 1

expected = ['dashboard','tenants','config','tools','rag','agents','abuse','voice','llm','audit']
missing = [p for p in expected if p not in pages]
if missing:
    print(f"ERROR: pages not found: {missing}")
    # Check what we found
    print(f"Found: {list(pages.keys())}")
    sys.exit(1)

# Build set of page line numbers
page_lines = set()
for p in pages.values():
    for i in range(p['start'], p['end'] + 1):
        page_lines.add(i)

# Write page files
os.makedirs(pages_dir, exist_ok=True)
for name in expected:
    p = pages[name]
    content = '\n'.join(lines[p['start']:p['end']+1])
    out_path = os.path.join(pages_dir, f'{name}.html')
    with open(out_path, 'w') as f:
        f.write(content)
    n = content.count('\n') + 1
    print(f"  -> pages/{name}.html ({n} lines)")

# Build skeleton: keep non-page lines, insert markers BEFORE the page region
skeleton_lines = []
for i, line in enumerate(lines):
    for name, p in pages.items():
        if p['start'] == i:
            skeleton_lines.append(f'<!--PAGE:{name}-->')

    if i not in page_lines:
        skeleton_lines.append(line)

skeleton = '\n'.join(skeleton_lines)

skel_count = skeleton.count('\n') + 1
orig_count = len(lines)
print(f"\nSkeleton: {skel_count} lines (was {orig_count})")
print(f"Saved: {orig_count - skel_count} lines")

marker_count = skeleton.count('<!--PAGE:')
page_file_count = len(os.listdir(pages_dir))
print(f"Markers: {marker_count}, page files: {page_file_count}")

with open(html_path, 'w') as f:
    f.write(skeleton)

print("Done!")
